FBX Importer

RevoluPowered opened the pull request:
This is the first prototype which we aim to release with Godot **3.2**

It needs extensive testing. We need your help to test it with as many models as possible.

Right now we have got a few issues to fix and patches will be put in place over the next few days.

**Remaining issues:**
- We do not currently support FBX SDK pivots (think Maya pivots) - we are working on implementing this but, this will need to be provided into a future update
- When an animated mesh has not executed any animation the rest pose is wrong.
- ~~Material code and culling need refactored and bugs with texture assignment need fixed.~~
- ~~We have not merged our experimental scaling system for application based scaling. We have submitted the PR for this already so don't worry! https://github.com/assimp/assimp/pull/2611~~
~~We still have to wait for merge of a few minor fixes into assimp to protect against import crashes and bugs
-- https://github.com/assimp/assimp/pull/2623~~

Any models which break can you please submit them to this address:
https://airtable.com/shr01fIPPqtrYoQfl
 
Would not have been possible without @kimkulling, @fire, @reduz, @hpvb, @akien-mga or IMVU INC.

Thank you for sponsoring us to do this work!

akien-mga commented on modules/assimp/import_utils.h:
```diff
@@ -0,0 +1,334 @@
+#ifndef IMPORT_UTILS_IMPORTER_ASSIMP_H
+#define IMPORT_UTILS_IMPORTER_ASSIMP_H
+
+#include "assimp/DefaultLogger.hpp"
```
These should use `<>` as they're thirdparty (potentially system-wide) headers, like `<string>` below.

RevoluPowered replied:
Resolved.

akien-mga commented on thirdparty/assimp/.gitignore:
```diff
@@ -0,0 +1,104 @@
+.idea
```
I don't think we need assimp's `.gitignore`.

RevoluPowered replied:
Fixed!

akien-mga commented on thirdparty/assimp/CREDITS:
```diff
@@ -1,183 +0,0 @@
-===============================================================
-Open Asset Import Library (Assimp)
```
On the other end the `CREDITS` were a nice touch so that Assimp contributors are properly credited for their work. It makes the "assimp team" from `LICENSE` explicit.

RevoluPowered replied:
I'll definitely re-add this

RevoluPowered replied:
Fixed

akien-mga commented on modules/assimp/editor_scene_importer_assimp.cpp:
```diff
@@ -28,6 +28,14 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 
+// todo: instead of creating a texture and discarding it / instead create an image and use it directly instead of pointlessly instancing
```
I'd suggest opening issues for each TODO instead, such comments often end up forgotten and never addressed (or addressed, but the comments themselves don't get removed).

RevoluPowered replied:
Fixed

akien-mga commented on modules/assimp/editor_scene_importer_assimp.cpp:
```diff
@@ -28,6 +28,14 @@
+*/
+// why? performance benefit 2x - directly fetching from visual server, instead of keeping the one we already have available
+
 #include "assimp/DefaultLogger.hpp"
```
These assimp includes should also use `#include <assimp/Importer.hpp>`.

RevoluPowered replied:
Fixed

akien-mga commented on thirdparty/assimp/include/assimp/irrXMLWrapper.h:
```diff
@@ -1,144 +0,0 @@
-/*
```
Makes sense to remove indeed.

@fire Not directly related to this PR and could be done later on, but I see lots of included paths added in `modules/assimp/SCsub` that we don't have in our distribution (like most upstream `contrib` stuff which we don't use nor ship).

akien-mga commented:
Thanks for the great work from everyone involved!

For merging, I'll need this rebased into something with a cleaner Git history, 60 commits is too much, especially since many of the commits are fixups of changes made in the same PR (which are not relevant to the Git history).

RevoluPowered commented:
@akien-mga will be working on the bugs first then I will rebase into a much smaller commits like 2-3. 

Just left it this way first so that if I need to fix anything by reverting anything I can.

We will update the code base with those changes, thanks though for reviewing it's going to make it much easier to get this merged now :)

RevoluPowered commented:
Application scaling is now implemented CM to M is now functional.

Some fixes need backported into assimp but this works.

RevoluPowered commented:
Most recent fixes added, this fix solves scaling issues with models.

We have now implemented the proper call to pass 'final scale' value properly to the ScaleProcess.

I have tested against the following models:
- CM, M models made in blender with different units (confirmed scale exactly the same)
- Log Cutter animated FBX
 
Screenshots of some models:

**Automatic scale working**

Log cutter
![image](https://user-images.githubusercontent.com/748770/63778053-0ee8aa00-c8dc-11e9-81f1-d1200909c9d6.png)

Bistro Interior model
![image](https://user-images.githubusercontent.com/748770/63778424-a817c080-c8dc-11e9-86cd-ab47ffc4300f.png)

Bistro Exterior model 

![image](https://user-images.githubusercontent.com/748770/63778503-d7c6c880-c8dc-11e9-9e34-5a29c2999f3c.png)

![image](https://user-images.githubusercontent.com/748770/63778526-e1e8c700-c8dc-11e9-9b1c-1b58644594ac.png)

As you can see CM to M is definitely fixed now. However I want to run more tests later this evening.

RevoluPowered committed:
Assimp FBX Import support

Issues fixed:
- Updated assimp to latest and backported fixes into godot.
- Fixed file scale being ignored from FBX file.
- Fixed bone removal
- Implemented proper armature binding
- Fixed recursion not always going through the entire path
- Implemented assimp global scaling system
- Fixed assimp global scale process to support unit conversion
- Implemented proper fbx scaling
- Fixed asserts caused by missing faces in some models which could crash
- Fixed valid bone removal
- Fixed root node being overwriten by assimp which caused data loss
- Fixed armature construction so that it works with multiple roots
- Implemented basic support for FBX standard materials
- Refactoring to improve code quality and improve function reuse.
- Simplified node creation from assimp scene into subsections: create_light, create_mesh, create_bone.
- Creating meshes is now done after hierarchy is created so that the skeleton is always available.
- Added support to assimp to support file scale in all formats which call SetFileScale.
- Many other fixes provided into assimp.

Known issues:
- FBX pivots from Maya do not currently work. (workaround: for now use blender import and export to remove pivot tracks)
- Hierarchy creates an extra node for each mesh - this was done intentionally but we intended to do a pass to remove these as they're a required node.
- When an animated mesh has not executed any animation the rest pose is wrong.

Co-authored-by: K. S. Ernest (iFire) Lee <ernest.lee@chibifire.com>

RevoluPowered commented:
Material code should be good for testing now. moving onto pivots then let's get this thing merged.

Geequlim commented:
Tested with a lot of FBX from the [SimpleTown Pack](https://syntystore.com/collections/simple-series/products/simple-town-cartoon-assets)

All model got an empty node with the name of the fbx file and the mesh was named as same with the file with a extention `2`.
![2019-09-01 17-58-41屏幕截图](https://user-images.githubusercontent.com/6964556/64074829-290ef780-cce3-11e9-88ec-ef1a933772b0.png)

The master branch doesn't have such empty node.

![2019-09-01 17-59-05屏幕截图](https://user-images.githubusercontent.com/6964556/64074822-1f858f80-cce3-11e9-8256-ee54224ff2e0.png)

Geequlim commented:
The folder structure are as below

```
Models
+------- ambo.fbx
Textures
+------- ambo.png
```
For the master branch: albedo textures of material referenced to the correct file path in the `Textures` folder as expected.

For this branch: Albedo textures are imported as builtin sub-resource of the generated materials instead of using the texture files in the project.

RevoluPowered commented:
> Tested with a lot of FBX from the [SimpleTown Pack](https://syntystore.com/collections/simple-series/products/simple-town-cartoon-assets)
> 
> All model got an empty node with the name of the fbx file and the mesh was named as same with the file with a extention `2`.
> ![2019-09-01 17-58-41屏幕截图](https://user-images.githubusercontent.com/6964556/64074829-290ef780-cce3-11e9-88ec-ef1a933772b0.png)
> 
> The master branch doesn't have such empty node.
> 
> ![2019-09-01 17-59-05屏幕截图](https://user-images.githubusercontent.com/6964556/64074822-1f858f80-cce3-11e9-8256-ee54224ff2e0.png)

Hi, @Geequlim thanks for reporting this issue I will make a patch on Monday to fix this.

> For this branch: Albedo textures are imported as builtin sub-resource of the generated materials instead of using the texture files in the project.

Do you not expect this to be the default behavior, as the mesh has the materials embedded it makes more sense to extract them to me, if you don't want it to do that you can disable it in the modelling export settings in blender so that it doesn't pack the images into the fbx?

akien-mga commented on modules/assimp/import_state.h:
```diff
@@ -0,0 +1,115 @@
+/*************************************************************************/
+/*  import_state.h                                       */
```
The trailing `*/` is misaligned.

akien-mga commented on modules/assimp/import_state.h:
```diff
@@ -0,0 +1,115 @@
+#include "scene/resources/animation.h"
+#include "scene/resources/surface_tool.h"
+
+#include "assimp/DefaultLogger.hpp"
```
Those should also use `< >` include style.

akien-mga commented on modules/assimp/editor_scene_importer_assimp.cpp:
```diff
@@ -28,24 +28,25 @@
-#include "assimp/pbrmaterial.h"
-#include "assimp/postprocess.h"
-#include "assimp/scene.h"
+#include <assimp/SceneCombiner.h>
```
Not a big deal, but as per our style guide these should come after Godot's own headers.

`#include "editor_scene_importer_assimp.h"` should also be the first incldue.

http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes

akien-mga replied:
"zutil.h" below is also a thirdparty header btw, should have the same treatment.

akien-mga commented on modules/assimp/editor_scene_importer_assimp.h:
```diff
@@ -44,60 +44,31 @@
+#include <assimp/LogStream.hpp>
+#include <assimp/Logger.hpp>
+
+#include "import_state.h"
```
These should be further up, thirdparty headers come last.

akien-mga commented on modules/assimp/import_state.h:
```diff
@@ -0,0 +1,115 @@
+};
+} // namespace AssimpImporter
+
+#endif // EDITOR_SCENE_IMPORT_STATE_H
```
Missing newline at EOF.

akien-mga commented on modules/assimp/import_utils.h:
```diff
@@ -0,0 +1,484 @@
+#ifndef IMPORT_UTILS_IMPORTER_ASSIMP_H
+#define IMPORT_UTILS_IMPORTER_ASSIMP_H
+
+#include <assimp/SceneCombiner.h>
```
Thirdparty headers come last (`<string>` too).

akien-mga commented on modules/assimp/import_utils.h:
```diff
@@ -0,0 +1,484 @@
+	}
+
+	/**
+     * Get global transform for the current node - so we can use world space rather than
```
Indentation issue.

akien-mga replied:
Same issue in many comments btw.

RevoluPowered replied:
Fixed in all thanks! will commit soon.

akien-mga commented on modules/assimp/import_utils.h:
```diff
@@ -0,0 +1,484 @@
+		return Ref<Image>();
+	}
+
+	// static bool GetAssimpTextureProperty(
```
Is this going to be used in the future, or is this dead code? Would be better to remove it if it's not used.

RevoluPowered replied:
Removed!

akien-mga commented on modules/assimp/import_utils.h:
```diff
@@ -0,0 +1,484 @@
+	}
+};
+
+#endif // IMPORT_UTILS_IMPORTER_ASSIMP_H
```
Missing newline.

akien-mga commented:
Could you amend the commit log to be more explicit about what it does? (The title is fine, but the body of the message could use more explanations before the "Co-authored-by:" line.)
See https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#format-your-commit-logs-with-readability-in-mind

fire commented on modules/assimp/editor_scene_importer_assimp.cpp:
```diff
@@ -122,22 +111,23 @@ Node *EditorSceneImporterAssimp::import_scene(const String &p_path, uint32_t p_f
-								 aiProcess_SortByPType |
-								 aiProcess_FindInvalidData |
+								 //aiProcess_SortByPType |
+								 // aiProcess_FindInvalidData |
```
This allows invalid files even though they're broken according to Assimp. This may cause previously acceptable files to be rejected and etc.

RevoluPowered replied:
I have enabled aiProcessFindInvalid data but sort by p type is not possible to enable right now as it would need lots of testing.

fire replied:
Sounds good. The sort by ptype was to remove vertex colors.

fire commented on modules/assimp/editor_scene_importer_assimp.cpp:
```diff
@@ -122,22 +111,23 @@ Node *EditorSceneImporterAssimp::import_scene(const String &p_path, uint32_t p_f
 								 //aiProcess_DropNormals |
 								 //aiProcess_GenSmoothNormals |
-								 aiProcess_JoinIdenticalVertices |
+								 //aiProcess_JoinIdenticalVertices |
```
This is an optimization. We can turn this on later.

RevoluPowered replied:
I don't trust this option right now so will leave it off because I need to test against a bunch of models to know if it's any good.

fire commented on modules/assimp/editor_scene_importer_assimp.cpp:
```diff
@@ -146,11 +136,11 @@ Node *EditorSceneImporterAssimp::import_scene(const String &p_path, uint32_t p_f
 								 //aiProcess_OptimizeGraph |
 								 //aiProcess_Debone |
 								 aiProcess_EmbedTextures |
```
Didn't we say embed texture caused problems? I forget.

RevoluPowered replied:
Removed! :)

akien-mga commented:
Thanks a lot to everyone involved!
